### DISCUSSION TITLE
Using puppeteer with lighthouse

### DISCUSSION THREAD

**Discussion by ha-douken** [GoogleChrome/lighthouse#12895] [Q&A]
You have an example of how to use puppeteer with lighthouse here:
_https://github.com/GoogleChrome/lighthouse/blob/master/docs/puppeteer.md_

In this example lighthouse will "open url" and puppeteer will add a style tag. I wonder if this will have any effect on the lighthouse result itself. Two thoughts:
- Is adding the style done before anything is painted or will lighthouse see styles changing?
- Is adding the style delaying further pocessing?

Probably this is not much of an issue regarding the example. But how about using puppeteers’s _page.setContent(html[, options])_ to e.g. test an optimized version of the html.

To give you more context I am looking for a way to intercept the HTML that is tested. I see that Chrome DevTools Protocol has _Network.getResponseBodyForInterception_ but it is experimental and not supported by puppeteer.

**Comment by patrickhulce** [comment]
> Is adding the style done before anything is painted or will lighthouse see styles changing?

It depends. Our example of puppeteer you link to is non-deterministic. `targetchanged` happens *very* early in page lifecycle but this is an inherently racy approach that might not register the style until after the page has painted. Lighthouse might observe a flash of paint before the style has been added, but the bulk of analysis waits for the entire page load and will primarily analyze the page with the stylesheet. 

> Is adding the style delaying further pocessing?

Only insomuch as processing that extra stylesheet if it were part of the page itself delays processing.

> But how about using puppeteers’s page.setContent(html[, options]) to e.g. test an optimized version of the html.

Because you'd then be analyzing two completely different pages in the same run. The test would be tainted by the resources and partial load of the original page. If this is what you want to accomplish, I highly recommend just creating a new URL to host that optimized content version (proxy server if you must).

#11313 is an effort to bring much tighter integrations with puppeteer, but even after that work is complete, racily attempting to replace the entire page content will still be ill advised.

**Comment by ha-douken** [comment]
Yes, a reverse proxy is my current solution.

Thanks for your help.
